Engine/VariableAnalysisBase.cs

```diff
@@ -1162,11 +1258,321 @@ public AssignmentTarget(AssignmentStatementAst ast)
+                {
+                    Type expressionType = (memberAst.Expression is TypeExpressionAst) ? (memberAst.Expression as TypeExpressionAst).TypeName.GetReflectionType() : null;
+
+                    if (expressionType != null)
+                    {
+                        result = GetPropertyOrFieldTypeFromMemberExpressionAst(expressionType, fieldName);
+                    }
+                    else
+                    {
+                        // check for class type
+                        TypeDefinitionAst psClass = FindClass(memberAst, (memberAst.Expression as TypeExpressionAst).TypeName.FullName);
+
+                        if (psClass != null)
+                        {
+                            MemberAst memAst = psClass.Members.FirstOrDefault(item => String.Equals(item.Name, fieldName, StringComparison.OrdinalIgnoreCase));
+
+                            if (memAst != null && memAst is PropertyMemberAst && (memAst as PropertyMemberAst).IsStatic)
+                            {
+                                result = (memAst as PropertyMemberAst).PropertyType.TypeName.GetReflectionType();
+                            }
+                        }
+                    }
+                }
+
+                #endregion
+            }
+
+            if (result != null)
+            {
+                return result;
+            }
+
+            #region RHS contains $this variable
+
+            // For the case where variable is $this
+            if (memberAst.Expression is VariableExpressionAst
+                && String.Equals((memberAst.Expression as VariableExpressionAst).VariablePath.UserPath, "this", StringComparison.OrdinalIgnoreCase))
+            {
+                // Check that we are in a class
+                TypeDefinitionAst psClass = FindClassAncestor(memberAst);
+
+                // Is static is false for this case
+                result = GetTypeFromClass(psClass, memberAst);
+            }
+
+            return result;
+            #endregion
+        }
+
+        /// <summary>
+        /// Get the type from an invoke member expression ast
+        /// </summary>
+        /// <param name="type"></param>
+        /// <param name="imeAst"></param>
+        /// <param name="methodName"></param>
+        /// <param name="isStatic"></param>
+        /// <returns></returns>
+        internal static Type GetTypeFromInvokeMemberAst(Type type, InvokeMemberExpressionAst imeAst, string methodName, bool isStatic)
+        {
+            Type result = null;
+
+            MethodInfo[] methods = (isStatic) ? type.GetMethods(BindingFlags.Public | BindingFlags.Static) : type.GetMethods(BindingFlags.Public | BindingFlags.Instance);
+            int argCounts = (imeAst.Arguments != null) ? imeAst.Arguments.Count : 0;
+
+            MethodInfo[] possibleMethods = methods.Where(method => String.Equals(method.Name, methodName, StringComparison.OrdinalIgnoreCase)
```

Comment:
Yes we don't care about the types of the arguments here. Because it is also hard to determine the types of the arguments so I'm doing a conservative approach, i.e I will only assign a type if all the methods with the same number of arguments have the same return type.